### DISCUSSION TITLE
Unexpected behaviour for asynchronous doSend-Call

### DISCUSSION THREAD

**Discussion by agering-hmmh** [spring-projects/spring-kafka#2889] [Ideas]
_doSend() returns CompletableFuture_

**Expectation**: sending is async and all errors are handled as CompletableFuture.exceptionally()

**Reality**:
- the Call is synchronous, Timeout blocks the application (caller needs to use `@Async`),
- Timeout causes a runtime exception outside of CompletableFuture, so the caller needs a `try catch` to handle all possible exceptions, otherwise where is status 500, although CompletableFuture.exceptionally defined

**Comment by artembilan** [comment]
This is how `KafkaProducer.send()` works. It is blocked on the `waitOnMetadata()`:
```
          ClusterAndWaitTime clusterAndWaitTime;
            try {
                clusterAndWaitTime = waitOnMetadata(record.topic(), record.partition(), nowMs, maxBlockTimeMs);
            } catch (KafkaException e) {
                if (metadata.isClosed())
                    throw new KafkaException("Producer closed while send in progress", e);
                throw e;
            }
```
That is not what we can control some way on `KafkaTemplate`.
